openstreetmap-carto.lua

```diff
@@ -303,134 +574,178 @@ function filter_tags_generic(tags)
         end
     end
 
-   -- Filter out objects that have no tags after deleting
-    if next(tags) == nil then
-        return 1, {}
+    return next(tags) == nil
+end
+
+--- Splits a tag into tags and hstore tags
+-- @return columns, hstore tags
+function split_tags(tags, tag_map)
+    local cols = {tags = {}}
+    for key, value in pairs(tags) do
+        if tag_map[key] then
+            cols[key] = value
+        else
+            cols.tags[key] = value
+        end
     end
+    return cols
+end
+
+phase2_admin_ways = {}
 
-    -- Convert layer to an integer
-    tags['layer'] = layer(tags['layer'])
-    return 0, tags
+-- TODO: Make add_* take object, not object.tags
+function add_point(tags)
+    local cols = split_tags(tags, columns_map.point)
+    cols['layer'] = layer(tags['layer'])
+    tables.point:add_row(cols)
 end
 
--- Filtering on nodes
-function filter_tags_node (keyvalues, numberofkeys)
-    return filter_tags_generic(keyvalues)
+function add_line(tags)
+    local cols = split_tags(tags, columns_map.line)
+    cols['layer'] = layer(tags['layer'])
+    cols['z_order'] = z_order(tags)
+    cols.tags['layer'] = nil
+    cols.way = { create = 'line', split_at = 100000 }
+    tables.line:add_row(cols)
 end
 
--- Filtering on relations
-function filter_basic_tags_rel (keyvalues, numberofkeys)
-    -- Filter out objects that are filtered out by filter_tags_generic
-    local filter, keyvalues = filter_tags_generic(keyvalues)
-    if filter == 1 then
-        return 1, keyvalues
-    end
+function add_transport_line(tags)
+    local cols = split_tags(tags, columns_map.line)
+    cols['layer'] = layer(tags['layer'])
+    cols['z_order'] = z_order(tags)
+    cols.tags['layer'] = nil
+    cols.way = { create = 'line', split_at = 100000 }
+    tables.transport_line:add_row(cols)
+end
 
-    -- Filter out all relations except route, multipolygon and boundary relations
-    if ((keyvalues["type"] ~= "route") and (keyvalues["type"] ~= "multipolygon") and (keyvalues["type"] ~= "boundary")) then
-        return 1, keyvalues
-    end
+function add_roads(tags)
+    local cols = split_tags(tags, columns_map.roads)
+    cols['layer'] = layer(tags['layer'])
+    cols['z_order'] = z_order(tags)
+    cols.tags['layer'] = nil
+    cols.way = { create = 'line', split_at = 100000 }
+    tables.roads:add_row(cols)
+end
 
-    return 0, keyvalues
+function add_polygon(tags)
+    local cols = split_tags(tags, columns_map.polygon)
+    cols['layer'] = layer(tags['layer'])
+    cols['z_order'] = z_order(tags)
+    cols.way = { create = 'area', split_at = nil }
+    tables.polygon:add_row(cols)
 end
 
--- Filtering on ways
-function filter_tags_way (keyvalues, numberofkeys)
-    local filter = 0  -- Will object be filtered out?
-    local polygon = 0 -- Will object be treated as polygon?
+function add_transport_polygon(tags)
+    local cols = split_tags(tags, columns_map.polygon)
```

Comment:
Shouldn't this be 

`     local cols = split_tags(tags, columns_map.transport_polygon)`

Reply:
Yes.

Reply:
I now tested this change in my own version of the style, and that seems to run fine, and the associated table is properly created.